proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,509 @@
+* Proposal: [0004](INF-0004-register-types-and-diagnostics.md)
+* Author(s): [Joshua Batista](https://github.com/bob80905)
+* Sponsor: TBD
+* Status: **Under Consideration**
+* Impacted Project(s): (LLVM)
+* PRs: [#87578](https://github.com/llvm/llvm-project/pull/87578)
+* Issues: [#57886](https://github.com/llvm/llvm-project/issues/57886)
+
+## Introduction
+Resources are often bound to registers of specific register types in HLSL.
+For example:
+```
+RWBuffer<float> rwbuf: register(u0);
+```
+Here we have the resource type (or variable type), `RWBuffer` with a resource 
+element type of `float` being declared as the variable `rwbuf`, and the variable 
+is bound to a register. The register type is `u` and the register number is `0`.
+There's a specific set of valid register types that can be used to bind a given 
+resource in HLSL. When invalid register types are used to bind resources, helpful 
+diagnostics need to be emitted so that the user knows why the resource type is 
+invalid, or why the register type isn't correct. This spec defines the behavior
+of the compiler for whenever an invalid register type is given, or for when a valid
+register type is used to bind a resource type that for whatever reason is 
+incompatible with that register type.
+
+## Motivation
+
+There are several cases in DXC in which diagnostics that are related to
+invalid register types are out of date or invalid. For example, in the case of:
+
+`float b : register(u4);`
+
+an error will be emitted recommending the use of the 'b, c, or i' register
+type. However the 'i' register type is no longer in support, and the 'b' 
+register type is only reserved for resource types that are constant buffers.
+It is worth noting that there is an overloading of the register(...) keyword
+using the 'c' register type to indicate an offset into a constant buffer for
+numeric types only, as opposed to specifying a resource binding.
+Additionally, it is possible the user is unaware that this variable won't 
+actually be used as a resource, but the compiler doesn't communicate that 
+to the user. It would be great for any HLSL developer to immediately have
+answers to the following questions:
+For any valid variable type that needs to be bound as a resource,
+what is the set of valid register types for that variable,
+and what are all the possible diagnostics that can be emitted if the variable
+is given a register type that isn't in the set of valid register types?
+The design below aims to specify the answer to these questions.
+
+## Proposed solution
+Firstly, for future reference, the set of relevant diagnostics that would be emitted
+depending on how the variable type and register type interact should
+be named and defined. Below is a table that lists out all the proposed diagnostic
+names and diagnostic messages that are relevant in this spec:
+
+| Diagnostic Name | Diagnostic message | Description |

```

Comment:
There's a lot of different diagnostics here, and some of them seem extremely similar to me.

Also, I wonder about some of them...for example:

```
float f : register(x0); // error: invalid register type 'x' used; expected 't', 'u', 'b', or 's'

// attempt to fix:

float f : register(t0); // error: unsupported register binding 't' on variable of type 'float'
float f : register(u0); // error: unsupported register binding 'u' on variable of type 'float'
float f : register(b0); // error: unsupported register binding 'b' on variable of type 'float'
float f : register(s0); // error: unsupported register binding 's' on variable of type 'float'
```

So the first error gave me four expected register types, none of which work.

---

I think `err_hlsl_unsupported_register_type_and_resource_type`  should only be generated if one of these unsupported types is used on a UDT.  For all other cases we know _exactly_ what register type is valid and should specify it there.

Reply:
Thanks for the feedback.
I agree that `err_hlsl_unsupported_register_type_and_resource_type` should be left to the UDT case. For the Resource case, I can definitely change the diagnostic to specify the exact register type that is expected. But for Basic types I don't think the compiler should necessarily recommend using "c", because it is unknown what the user is trying to do. Recommending "c" is outside the scope of this diagnostic infrastructure because it has nothing to do with binding a resource. Updated the spec to reflect this.